docs: add a short guide on security. (#2201)
electric-sql/electric: website/docs/guides/security.md

/>

# Security

How to secure data access and [encrypt data](#encryption) with Electric.

## Data access

Electric is a [sync service](/product/electric) that runs in front of Postgres. It connects to a Postgres database using a [`DATABASE_URL`](/docs/api/config#database-url) and exposes the data in that database via an [HTTP API](/docs/api/http).

<figure>
  <a :href="ComponentsJPG">
    <img :src="ComponentsPNG" class="hidden-sm"
        alt="Illustration of the main components of a successfull deployment"
    />
    <img :src="ComponentsSmPNG" class="block-sm"
        style="max-width: 360px"
        alt="Illustration of the main components of a successfull deployment"
    />
  </a>
</figure>

This API is [public by default](#public-by-default). It should be secured in production using [network security](#network-security) and/or an [authorization proxy](#authorization).

### Public by default

Electric connects to Postgres as a normal [database user](https://www.postgresql.org/docs/current/user-manag.html). It then exposes access to **any&nbsp;data** that its database user can access in Postgres to **any&nbsp;client** that can connect to the Electric HTTP API.

> If you don't want to expose public access to the contents of your database, you **must** secure access to the Electric HTTP API.

Reviewer: ```suggestion
You generally do _not_ want to expose public access to the contents of your database, so you **must** secure access to the Electric HTTP API.
```

```hcl
@denied not remote_ip 100.200.30.40 100.200.30.41
abort @denied
```

This approach is useful when you're using Electric to sync into trusted infrastructure. However, it doesn't help when you're syncing data into client devices, like apps and web browsers. For those, you need to restrict access using an authorizing proxy.

### Authorization

Electric is designed to run behind an [authorizing proxy](/docs/guides/auth#requests-can-be-proxied).

This is the primary method for securing data access to clients and apps and is documented in detail, with examples, in the [Auth guide](/docs/guides/auth).

## Encryption

Electric syncs ciphertext as well as it syncs plaintext. You can encrypt and decrypt data in HTTP middleware or in the local client.

### End-to-end encryption

For example, you can achieve end-to-end encryption by:

- *encrypting* data before it leaves the client
- *decrypting* data when it comes off the replication stream into the client

You can see an example of this in the [encryption example](/demos/encryption):

<<< @../../examples/encryption/src/Example.tsx{tsx}

> ### Key management

Reviewer: I don't have a lot of experience with key distribution but isn't storing the key (unencrypted) alongside the encrypted data more-or-less make it basically the same as if it wasn't encrypted? Since anyone with access to the db can now decrypt whatever data they want?

Author: You could store keys in a different database and just sync the right ones. I do think that the point of Electric being good to sync keys is valid: it's just like the distributed config case. You need to share the right keys with the right users. What better mechanism?

Reviewer: 🤷 like I said I dunno — maybe if the key is encrypted before sharing so it's just a backup/cloud syncing mechanism.

Author: Have jigged the description a bit.

Author: Ultimately for e2e encryption you need to get keys into the client.

Reviewer: you can have client-generated keys right?

Author: Yup, but if you're sharing the data, you have to share keys.
